ups/toman/man/ups.1
.TH upd 1 1999-12-08
.SH NAME
ups, upd, setup, unsetup -- access to Fermilab local products
.SH DESCRIPTION

.nf
.B "setup [opts] product [version]"
 Makes product instance available for use.
 Options
     -0, -1, -2, -3, -?, -c, -d, -e, -f, -g, -H, -j, -k, -M, -m, -n,
     -O, -o, -P, -q, -R, -r, -s, -t, -U, -V, -v, -Z, -z

.B "unsetup [opts] product [version]"
 Makes the product no longer available for use by undoing changes made
 to the environment by setup.
 Options
     -0, -1, -2, -3, -?, -c, -d, -e, -f, -g, -H, -j, -M, -m, -n, -O,
     -o, -P, -q, -r, -s, -t, -U, -V, -v, -Z, -z

.B "upd fetch [options] [-J filename] product version"
 Retrieves a single file or directory maintained in a UPS distribution
 database, and downloads it to the user node, placing it relative to the
 current working directory. This command cannot retrieve files from
 within a tar file. In an archive distribution database, the only files
 provided in the appropriate format are the ups subdirectory files, the
 README and INSTALL_NOTE files, and the table and version files. If file
 name or path is not specified (via the -J option), the command returns
 a recursive list of directories and files that are available for
 retrieval. Warning: when using a live distribution database, this list
 may be very long. product name
 Options
     -0, -1, -2, -3, -?, -c, -d, -f, -g, -H, -h, -J, -n, -o, -q, -s,
     -t, -v, -z

.B "ups get -F [options] product [version]"
 ups get -F outputs a list of files that are maintained outside of the
 product area (exclusive of table files, for which the location is
 maintained in the version file). In a future release, ups get may
 acquire additional functions.
 Options
     -0, -1, -2, -3, -?, -c, -d, -F, -f, -g, -H, -M, -m, -n, -o, -q,
     -r, -t, -u, -v, -Z, -z

.B "upd get -F options product [version]"
 upd get -F outputs a list of files that are maintained outside of the
 product area (exclusive of table files, for which the location is
 maintained in the version file) on the distribution node and which must
 be distributed with the product. It is used similarly to ups get. In a
 future release, upd get may acquire additional functions.
 Options
     -0, -1, -2, -3, -?, -c, -d, -F, -f, -g, -H, -h, -M, -m, -n, -o,
     -q, -r, -t, -u, -v, -Z, -z

.B "upd install options product [version]"
 Retrieves a product instance and by default its UPS product
 requirements as needed from a UPS distribution node. It installs it (or
 them) on the user node (unwinding if transferred in tar format) and
 declares it/them to a local UPS database. The -G option is provided to
 allow you to pass UPS options to the local ups declare command which
 gets called by UPD. product name
 Options
     -0, -1, -2, -3, -?, -C, -c, -d, -f, -G, -g, -H, -h, -I, -i, -j,
     -M, -m, -n, -O, -o, -q, -R, -r, -s, -t, -U, -v, -z


.B "ups list [options] [product] [version]"
 Outputs product instance information contained in the ups databases in
 one of two formats; one designed to be read by a person and the other
 for piping information to another command.
 Options
     -0, -1, -2, -3, -?, -a, -c, -d, -f, -g, -H, -K, -l, -M, -m, -n,
     -o, -q, -r, -t, -u, -v, -Z, -z

.B "upd list [options] [product] [version]"
 Performs ups list on the distribution UPS database(s).
 Options
     -0, -1, -2, -3, -?, -a, -c, -d, -f, -g, -H, -h, -K, -l, -M, -m,
     -n, -o, -q, -r, -t, -u, -v, -Z, -z

.B "upd update component options product [version]"
 Retrieves either the table file or ups subdirectory (depending on the
 component specified) for a product instance, and by default for its UPS
 product requirements, from a UPS distribution database for the purpose
 of updating the pre-existing one(s) on a local node. Overwriting occurs
 only if the MODIFIED date in the (remote) version file that points to
 the retrieved component is more recent than that in the pre-existing
 local version file; file time-stamps are not used for comparison. The
 update will fail if there is no corresponding pre-existing component on
 the user node. component (specified as table_file or ups_dir ), and
 product name
 Options
     -0, -1, -?, -c, -d, -f, -G, -g, -H, -h, -I, -i, -j, -n, -O, -o,
     -P, -q, -R, -s, -t, -v, -z

.B "Developer and Installer Commands"

.B "ups depend [options] product [version]"
 Identifies UPS product requirements of the specified product or
 instance(s).
 Options
     -0, -1, -2, -3, -?, -c, -d, -f, -g, -H, -j, -K, -l, -M, -n, -o,
     -q, -R, -r, -s, -t, -u, -v, -Z, -z

.B "upd depend [options] product [version]"
 Identifies UPS product requirements of the specified product or
 instance(s) as declared in the distribution database. It is used
 similarly to ups depend.
 Options
     -0, -1, -2, -3, -?, -c, -d, -f, -g, -H, -h, -j, -K, -l, -M, -n,
     -o, -q, -R, -r, -s, -t, -u, -v, -Z, -z

.B "ups exist [options] product [version]"
 Determines if a setup command with the same options would likely
 succeed. Specifically, it determines whether ups can create the
 temporary file of actions that setup would perform. If so, it creates
 the file, but it does not execute it. This can be used as a quick test
 for existence of a particular product instance.
 Options
     -0, -1, -2, -3, -?, -c, -d, -e, -f, -g, -H, -j, -k, -M, -m, -n,
     -O, -o, -P, -q, -r, -t, -u, -V, -v, -Z, -z

.B "upd exist [options] product [version]"
 Determines if a setup command with the same options would likely
 succeed on the remote node. Specifically, it determines whether ups can
 create the temporary file of actions that setup would perform. If so,
 it creates the file, but it does not execute it. This can be used as a
 quick test for existence of a particular product instance. It is used
 similarly to
 ups exist productname
 Options
     -0, -1, -2, -3, -?, -c, -d, -e, -f, -g, -H, -h, -j, -k, -M, -m,
     -n, -O, -o, -P, -q, -r, -t, -u, -V, -v, -Z, -z

.B "upd addproduct [options] product version"
 Adds one or a combination of the following to the KITS database: a
 product tar file, an unwound product (must be declared to a local UPS
 database), a table file, or a ups directory. Required elements depend
 on the object being added.
 Options
     -0, -1, -2, -3, -?, -A, -c, -D, -d, -f, -g, -H, -h, -M, -m, -n,
     -O, -o, -P, -p, -q, -r, -s, -T, -t, -U, -V, -v, -z

.B "upd modproduct [-m tablefile] [-U upsdir] [-g chain] product version"
.B "-f flavor"
 Used to a chain to a product instance declared in KITS Also used to
 replace a table file or a ups directory for a product instance in KITS.
 Required elements depend on the object being modified.
 Options
     -0, -1, -2, -3, -?, -c, -d, -f, -g, -H, -h, -M, -m, -n, -o, -q,
     -s, -t, -U, -v, -z

.B "upd delproduct [opts] product version -f flavor"
 Deletes a specified product instance declaration and its associated tar
 file (if any) from the KITS database.
 Options
     -0, -1, -2, -?, -c, -d, -f, -g, -H, -h, -n, -o, -q, -s, -t, -v,
     -z

.B "upd repproduct [options] product version"
 Equivalent to a upd delproduct followed by a upd addproduct
 Options
     -0, -1, -2, -3, -?, -A, -c, -D, -d, -f, -g, -H, -h, -M, -m, -n,
     -O, -o, -P, -p, -q, -r, -s, -T, -t, -U, -V, -v, -z

.B "ups verify [options] [product] [version]"
 Performs checks of the information in all the database files for the
 specified instances to determine if there are any errors or
 inconsistencies.
 Options
     -0, -1, -2, -3, -?, -a, -c, -d, -f, -g, -H, -K, -l, -M, -m, -n,
     -o, -q, -r, -t, -u, -v, -Z, -z

.B "upd verify [options] [product] [version]"
 Performs ups verify on the distribution UPS database(s).
 Options
     -0, -1, -2, -3, -?, -a, -c, -d, -f, -g, -H, -h, -K, -l, -M, -m,
     -n, -o, -q, -r, -t, -u, -v, -Z, -z

.B "ups configure [options] product [version]"
 Runs the "configure" action of the product, usually to allow
 constructing symbolic links, copying files, or automatic local
 customization of the product.
 Options
     -0, -1, -2, -3, -?, -c, -d, -f, -g, -H, -M, -m, -n, -O, -o, -P,
     -q, -r, -t, -U, -V, -v, -Z, -z

\fBups copy [options] product [version] -G "[options] product version"\fP
 Designed as a ups product development tool allowing a new instance of a
 product to be declared "like" another.
 Options
     -0, -1, -2, -3, -?, -A, -b, -C, -c, -D, -d, -f, -G, -g, -H, -M,
     -m, -n, -O, -o, -p, -q, -r, -T, -t, -U, -u, -V, -v, -W, -X, -Z,
     -z

.B "ups declare [-r home] [-m table_file_name] [-M table_file_loc]"
.B "[-U ups_dir] [options] product version"
 Adds a new product instance and-or chain to a UPS database, or changes
 the chain information on an instance.
 Options
     -0, -1, -2, -3, -?, -A, -b, -C, -c, -D, -d, -f, -g, -H, -L, -M,
     -m, -n, -O, -o, -q, -r, -s, -T, -t, -U, -u, -V, -v, -Z, -z

.B "ups modify [options] product [version]"
 Allows user to manually edit product files and then performs
 syntax/content validation.
 Options
     -0, -1, -2, -3, -?, -A, -a, -c, -d, -E, -e, -f, -H, -K, -k, -M,
     -m, -N, -n, -o, -p, -q, -r, -s, -T, -U, -V, -v, -Z, -z

.B "ups start [options] product [version]"
 Allows automatic startup for products that need to load drivers start
 daemons or perform other actions required at system boot time.
 Options
     -0, -1, -2, -3, -?, -c, -d, -f, -g, -H, -M, -m, -n, -O, -o, -P,
     -q, -r, -s, -t, -U, -V, -v, -w, -Z, -z

.B "ups stop [options] product [version]"
 For products that get started automatically at system startup this
 allows automatic stop of products at system shutdown.
 Options
     -0, -1, -2, -3, -?, -c, -d, -f, -g, -H, -M, -m, -n, -O, -o, -P,
     -q, -r, -s, -t, -U, -V, -v, -Z, -z

.B "ups tailor [options] product [version]"
 Allows the entry of any once per installation information that requires
 user input; e.g. the location of specific hardware devices. If a
 product instance has tailor functions, the command must be run by the
 installer after the product is declared to the ups database.
 Options
     -0, -1, -2, -3, -?, -c, -d, -f, -g, -H, -h, -K, -M, -m, -n, -O,
     -o, -P, -q, -r, -s, -t, -U, -V, -v, -Z, -z

.B "ups touch [options] product [version]"
 Changes the MODIFIED date and MODIFIER to the current time and the
 current user, respectively.
 Options
     -0, -1, -2, -3, -?, -b, -c, -d, -f, -g, -H, -M, -m, -n, -o, -q,
     -t, -v, -Z, -z

.B "ups unconfigure [options] product [version]"
 This calls the products "unconfigure" action, which should undo
 whatever the products "configure" action does.
 Options
     -0, -1, -2, -3, -?, -c, -d, -e, -f, -g, -H, -K, -k, -M, -m, -n,
     -O, -o, -P, -q, -r, -s, -t, -U, -V, -v, -Z, -z

.B "ups undeclare [options] product [version]"
 Allows removal of instances from the ups database.
 Options
     -0, -1, -2, -3, -?, -C, -c, -d, -f, -g, -H, -K, -k, -M, -m, -n,
     -O, -o, -o, -P, -q, -r, -s, -t, -U, -V, -v, -Y, -y, -Z, -z

.B "Command Option Descriptions"

.B "-0"
 NULL flavor string; use in place of -f option
 Note: can also be used inside -G ""

.B "-1"
 OS flavor specification of distribution node up to the generic OS
 (e.g., SunOS); use in place of -f option
 Note: can also be used inside -G ""

.B "-2"
 OS flavor specification of distribution node up to the major release
 (e.g., SunOS+5); use in place of -f option
 Note: can also be used inside -G ""

.B "-3"
 most significant OS flavor specification of distribution node, or the
 full specification (e.g., SunOS+5.5); use in place of -f option
 Note: can also be used inside -G ""

.B "-?"
 print out usage information for command
 Note: C shell users, enclose -? in double quotes (e.g., ups declare
 "-?" ); -? is interpreted by sh.

.B "-A"
 nodes set authorized nodes to include only specified nodes
 Note: can also be used inside -G ""

.B "-a"
 include all instances found in searched distribution database(s)

.B "-b compile_file"
 Specify the name of a file to hold compiled functions.

.B "-c"
 Don't run configure/unconfigure current/uncurrent, etc. actions
 Note: can also be used inside -G ""

.B "-c"
 current chain for product instance in distribution database
 Note: can also be used inside -G ""

.B "-D URL"
 specify product's master source location, (informational only)
 Note: can also be used inside -G ""

.B "-d"
 specifies the development chain for product instance in distribution
 database, Equivalent to "-g development"
 Note: can also be used inside -G ""

.B "-E"
 Run the editor specified by $EDITOR on the requested file
 Note: can also be used inside -G ""

.B "-e"
 sets $UPS_EXTENDED to the value 1 when performing actions. (This may
 have an effect if the table file uses the variable)

.B "-F"
 return list of files (excluding table file, for which the location is
 maintained in the version file) that are associated with the product on
 the distribution node but that are maintained external to the product's
 root directory and external to the database

.B "-f"
 flavor flavor of product on distribution node; for products with UPS
 product requirements, this applies only to the main product.
 Notes: Can also be used inside -G "". See note under -H.

\fB-G "options_to_pass_to_local_ups_declare_command"\fP
 options to be modified over the existing command line options,
 and then to be given in a ups declare command.
 Note: If you specify -G, upd install does two declarations, one
 without the options specified with -G overridden, and a second one with
 them.

.B "-g chainName"
 create or use product instance associated with chainName.
 Note: can also be used inside -G ""

.B "-H localFlavor"
 specify host flavor to use in order to build a flavor list for the
 product and all of its requirements
 Note: this differs from -f in that it specifies a group of flavors, not
 a specific one.

.B "-h"
 hostName host name (node name) that contains the distribution UPS
 database; the default is fnkits.fnal.gov. Usually just the plain host
 name is required, however the -h option can always specify any of the
 following:
    o a plain host name (e.g., fred.sub.domain)
    o a host and Webserver port number (e.g., fred.sub.domain:8080 )
    o a full URL to the ups.cgi cgi script (e.g.,
      http://fred.sub.domain:8080/cgi-bin/some/dir/ups.cgi )
    o a file url to a locally mounted products database (e.g. a cdrom)
      like file://localhost/mnt/cdrom/ups/db

.B "-I"
 unwind product tar file on the local node after transmission (default
 is to unwind during transmission); this option allows ftp to re-get a
 tar file if your network connection is lost during the transfer.

.B "-i"
 ignore errors; if a UPS product requirement can't be distributed,
 continue

.B "-J"
 filename individual file to fetch (for archive database, must be one of
 INSTALL_NOTE, README, or a version or table file)

.B "-j"
 setup/get only the specified (top level) product (ignore UPS product
 requirements)

.B "-K"
 "keywords" return information for the specified keywords only, in
 machine-readable format.

.B "-k"
 prevents execution of unsetup files prior to (subsequent) setup.

.B "-L"
 add the STATISTICS keyword for this instance in its version file.

.B "-l"
 produce a full (long) listing

.B "-M tableFileDir"
 table file directory
 Note: can also be used inside -G ""

.B "-m tableFile"
 table file name
 Note: can also be used inside -G ""

.B "-N filename"
 specify a particular file to be checked by ups verify
 Note: can also be used inside -G ""

.B "-n"
 use/create new chain (i.e. chain name new ) for product instance
 Note: can also be used inside -G ""

.B "-O"
 "flags" used to specify $UPS_OPTIONS value to pass to actions.
 Note: can also be used inside -G "" to set $UPS_OPTIONS to "options"

.B "-o"
 use/create old chain (i.e. chain name old ) for product instance
 Note: can also be used inside -G ""

.B "-P"
 Do not look in the ups database for this product instance, instead rely
 only on those items specified on the command line.

\fB-p "description"\fP
 product description

.B "-q [qual1[:qual2...]][?opt1[:?opt2...]]"
 product instance qualifiers; multiple qualifiers are separated by
 colons, optional ones have question marks.
 Note: can also be used inside -G ""

.B "-R"
 use only the required UPS product requirements (listed in table file
 using setupRequired ); ignore the optional ones.

.B "-r home"
 Specify product's root directory.

.B "-s"
 list what command would do; but do not execute the command

.B "-T URL"
 Specify URL or pathname for archive file.

.B "-t"
 test chain (i.e. chain name test ) for product instance in distribution
 database
 Note: can also be used inside -G ""

.B "-U upsDir"
 ups directory location

.B "-u dir"
 directory for compiled functions files (see also -b)

.B "-V"
 Do not delete temporary script files or partially installed products
 when command finishes, instead list them.

.B "-v"
 print out extra `debugging' information. To get progressively more
 information, use multiple v 's, e.g., -vv, -vvv (up to four).

.B "-W"
 Use environment variables that show what is currently set up to
 determine instances of UPS product requirements.

.B "-w"
 On ups start, stop the product first, then start it.

.B "-X"
 echo the command to execute, but don't run it.

.B "-Y"
 delete product root directory and all files under it, without asking.

.B "-y"
 delete product root directory and all files under it, after asking.

.B "-Z"
 Time this command.

.B "-z"
 database Used to specify the distribution database; whole path can be
 specified here. The default value depends on the UPD configuration on
 the distribution node and the value of $PRODUCTS there.
 Note: can also be used inside -G ""
.fi
.SH SEE ALSO
sh(1), csh(1), bash(1), tcsh(1)
.SH BUGS
 User defined variables in group stanzas get lost.
 A setupRequired using -g current and -H flavor may coredump.
 The bash shell complains about sourcing /dev/null.


doc/RELEASE_NOTES
Changes made in the v4_5 release of ups:

Added tests of the -C option to inhibit executing unchain actions in the
declare and undeclare routines.  Reworked the code some to emit -v messages
when -C causes actions to be skipped.  Added a "free" to match the "malloc"
for the unchain in ups_declare.c

Added code to override the -j option setting saved in the SETUP_prod
environment variable when unsetting up a product either directly or
indirectly by setting up a previously set up product.  This prevents a
dependent product's getting set up twice without an intervening unsetup in
situations like the following:  Product A has Product B as a dependent
product.  Product B is setup.  Product A is setup with the -j option.
Setting up Product A unsets up A with the -j from the SETUP_A environment
variable, failing to unsetup B, but going on to set it up again.

Made the upsutl_finish_up routine return an error status if a problem
occurs, most likely in calls to the upsfil_flush routine.  This prevents ups
from exiting with a success code when a declare fails, due to file
permission problems for example, which in turn confuses upd.

Mapped the appropriate [un]declare action's option list to the various
[un]chain actions in the get_ugo routine for its call to upsugo_bldcmd, and
fixed the upsugo_next routine crash if we get null for the valid_option
list.  These changes allow table files to have setupRequired/Optional
actions in chain stanzas (current, development, etc.) and not have ups crash
when called with -H, as upd is wont to do.

Changed the upsget_remall and upsget_allout routines in upsget.c to use the
g_var_subs table to write environment variables to the temporary file.  The
initial motivation for this was to make a generic fix to crashes caused by
attempts to use UPS_ENV in actions when no database is available, and hence
some of the environment variables have null values.  The other motivation is
that is makes it easier to add items to the list to be written to the
temporary file, and eliminates some redundant code.  The upsget_allout and
upsget_envout routines now use apostrophes for quoting values, rather than
quote marks, which fixes the problem when defining SETUP_<something> with an
-O "some options" string embedded in it.  All environment variables are now
quoted, and the formatting is a little easier to look at.

Fixed the cat/man page copy code in upscpy.c to not get confused by dots
occuring in directories or in product names (like kadmin.local, e.g.).
Fixed the code to properly handle .z, .gz, and .Z extensions.  It now also
accepts most any subsection letter (but not g, z, or Z).

Moved the UPSCPY informational messages from level 0 to level 1, meaning
that the user will have to use -v with declare to see the long list of
"UPSCPY: Copying man files" and similar messages.  This change has the happy
side-effect of clearing up much clutter in the testing process.

Defined the UPS_UPS_DIR variable if UPS_ENV is specified, in upsget.c.

Fixed the handling of the DB alias for the DATABASE keyword, so that ups
list -KDB works.

Enabled the -R option for the setup command.

Added a ups(1) man page.

Changed the command index passed to upsutl_finish_up by main to e_help if
help is requested.  This prevents a spurious /dev/null from being echoed for
ups -?, ups setup -?, and ups unsetup -? commands.

Option list clean-up items in upsact.c:
  - Removed dead options A:b:CD:O:p:r:T:u:U:V from the list for touch.
  - Removed all instances of the B: option, for now at least.
  - Removed the : from the e options.
  - Removed all instances of the h: option.
  - Removed options K:l from the list for verify.
  - Removed misleading options m:M: from the list for touch.
  - Removed option o from the list for modify.
  - Removed option V from the lists for declare, get, list and verify.
  - Removed option x, which will never be implemented.
  - Removed most items from the option list for help.
  - Made the ordering of the option lists consistent.


Test case clean-up and other changes:
  - Changed the get (get3.in, get4.in, relchk.in) and action routine
    (fly1.in and fly2.in) test input files to reflect the changes
    in the quoting of environment variables in upsget.c.
  - Made the /dev/null output in the action (oregano.in) test disappear.
  - Added numerous real and bogus cat/man files (spot) to copy and ignore.
  - Doctored the declare test list (tostuff.in) to match the newly added
    cat/man files.
  - Made the blank line output in the depend test disappear (bee4.in).
  - Returned to using the lp account for tilde expansion checks in get
    (get1.in and get2.in).
  - Added UPS_UPS_DIR to get test output (get3.in, get4.in, relchk.in).
  - Added code to sort the undeclare action list (spot3.in), since it
    depends on file system order otherwise.

Eliminated the old [un]setup.[c]sh scripts.

Code clean-up items to eliminate compiler warnings:
  - Large number of void casts added to printf's, strcpy's, etc.
  - Added zero comparisons in some if, while, and ?: statements
  - Added the macro SHUTUP and routine shutup to pretend to use
    standardized parameters
  - Eliminated some empty else clauses
  - Moved some return statements to the bottom of the routine
  - Removed some unused parameters
  - Fixed the widening of some unsigned chars
  - Defined some missing prototypes, fix some function declarations
  - Added initializers to some declarations
  - Changed some NULL initializers to 0s
  - Fixed some "unreachable" breaks in switches

Makefile clean-up items:
  - Added man and ups to the clean and spotless actions in the main Makefile
  - Merged spotless and clean targets, where they are the same
  - Made spotless depend on clean targets, where they differ
  - Cleaned up lots of files so that 'make spotless' returns the directory
    tree to its initial state, in the doc, man, and test Makefiles.


ups/ups.py
import os
import re
import sys
import string

DEFAULT_PYTHON_VERSION = ''
if os.environ.has_key('SETUPS_DIR'):
    DEFAULT_SETUPS_DIR = os.environ['SETUPS_DIR']
else:
    DEFAULT_SETUPS_DIR = '/usr/local/etc'

#############################################################
# User interface:
#  import ups
#  optional:
#       ups.getUps( [setupsDir=/path/to/setups.sh] )
#  
#  ups.setup('my project spec' [, setupsDir=/path/to/setups.sh])
#############################################################
# returns the address of a singleton upsManager object:

singletonUps = None
def getUps(setupsDir=DEFAULT_SETUPS_DIR):
    global singletonUps
    if ( singletonUps is None ):
        singletonUps = upsManager(setupsDir)
    return singletonUps

def setup(arg, setupsDir=DEFAULT_SETUPS_DIR):
    upsmgr = getUps()
    upsmgr.setup(arg)

#############################################################

def set_setupsDir(setupsDir=DEFAULT_SETUPS_DIR):
    global DEFAULT_SETUPS_DIR
    DEFAULT_SETUPS_DIR = setupsDir
#############################################################

# force a reload and exec of the requested version of
# python:
def use_python(pythonVersion=DEFAULT_PYTHON_VERSION):
    if ( pythonVersion != DEFAULT_PYTHON_VERSION ):
        vm = '.*'
    else:
        vm = pythonVersion

    obj = getUps()

    # are we already using the requested version?
    alreadyUsing = ( os.environ.has_key('SETUP_PYTHON') and
                     re.search('python ' + pythonVersion, os.environ['SETUP_PYTHON']) != None )

    if ( not alreadyUsing ):
        if os.environ.has_key('SETUP_PYTHON'):
            obj._inhaleresults(os.environ["UPS_DIR"] + '/bin/ups unsetup python')
        obj._inhaleresults(os.environ["UPS_DIR"] + '/bin/ups setup python ' + pythonVersion)

        # were we running the python interpreter itself?  Special handling required!
        if ( sys.argv == [''] ):
            sys.argv = [ os.environ['PYTHON_DIR'] + '/bin/python' ]
        else:
            sys.argv.insert(0, os.environ['PYTHON_DIR'] + '/bin/python')

        # bye bye, exec into another context:
        os.execve( os.environ['PYTHON_DIR']+'/bin/python', sys.argv, os.environ )

##############################################################################

class upsException(Exception):
    def __init__(self, msg):
        self.args = msg
    
class upsManager:
    def __init__(self, setupsDir=DEFAULT_SETUPS_DIR):

        # initial setup of ups itself:
        os.environ['UPS_SHELL'] = 'sh'
	f = os.popen( '. %s/setups.sh; ' % setupsDir + \
                      'echo os.environ\\[\\"UPS_DIR\\"\\]=\\"$UPS_DIR\\"; ' + \
                      'echo os.environ\\[\\"PRODUCTS\\"\\]=\\"$PRODUCTS\\";' + \
                      'echo os.environ\\[\\"SETUP_UPS\\"\\]=\\"$SETUP_UPS\\"')
	exec f.read()
	f.close()

        # we need to initialize the following so that we can
        #  make the correct changes to sys.path later when products
        #  we setup modify PYTHONPATH
        try:
            self._pythonPath = os.environ['PYTHONPATH']
        except KeyError:
            self._pythonPath = ''
        self._sysPath = sys.path
        (self._internalSysPathPrepend, self._internalSysPathAppend) = self._getInitialSyspathElements()

    ############################################################################
    # set a product up:
    def setup(self, arg):
        try:
            self._inhaleresults(os.environ["UPS_DIR"] + '/bin/ups setup ' + arg)
        except upsException:
            raise
        self._updateImportPath()

    ############################################################################
    # unsetup a product:
    def unsetup(self, arg):
        try:
            self._inhaleresults(os.environ["UPS_DIR"] + '/bin/ups unsetup ' + arg)
        except upsException:
            raise
        self._updateImportPath()

    ############################################################################ 
    # PRIVATE METHODS BELOW THIS POINT.
    #
    def _getInitialSyspathElements(self):
        pyPath = string.split(self._pythonPath, ':')
        sysPath = self._sysPath

        # sysPath always includes '' at the front of the list
        #  (current working directory)
        internalSysPathPrepend = ['',]
        sysPath = sysPath[1:]

        # now, what else is in sysPath that is NOT in PYTHONPATH
        # (and hence must have come from the internals)?
        internalSysPathAppend = []
        for element in sysPath:
            if ( element not in pyPath ):
                internalSysPathAppend.append(element)
        return (internalSysPathPrepend, internalSysPathAppend)
        
    ############################################################################
    def _setPythonPath(self):
        self._pythonPath = os.environ['PYTHONPATH']

    ############################################################################
    def _updateImportPath(self):
        # make sure that sys.path reflects what is in PYTHONPATH after
        # any setups are performed!
        if (os.environ['PYTHONPATH'] != self._pythonPath):
            pypathElements = string.split( os.environ['PYTHONPATH'], ':' )
            # sys.path includes the current working directory as the first element:
            sys.path = self._internalSysPathPrepend
            # now append each pythonpath element
            for element in pypathElements:
                sys.path.append(element)
            # now append whatever was 'internal' originally
            sys.path = sys.path + self._internalSysPathAppend
        # update our reference to PYTHONPATH
        self._setPythonPath()

    ############################################################################
    def _inhaleresults(self, cmd):
	(stdin,stdout,stderr) = os.popen3(cmd)
        filename = stdout.read()
	filename = filename[0:-1]
        if ( filename == "/dev/null" ):
            msg = stderr.read()
            raise upsException(msg)
        stdin.close()
        stdout.close()
        stderr.close()

	setup = open(filename,"a")
	setup.write(""" 
echo "--------------cut here-------------"
cat <<'EOF' | python
import os
import re
for v in os.environ.keys():
  fix= re.sub( "\'", "\\'", os.environ[v])
  print "os.environ['"+v+"'] = '" + fix + "'" 
EOF
"""
        )
	setup.close()

	f = os.popen("/bin/sh " + filename)
	c1 = f.read()
	f.close()
	c1 = re.sub( '.*--------------cut here-------------', '', c1)
	os.environ = {}
	exec c1




